compiler/rustc_mir_build/src/thir/pattern/_match.rs

```diff
@@ -2207,19 +2111,7 @@ fn pat_constructor<'tcx>(
             if let Some(int_range) = IntRange::from_const(tcx, param_env, value, pat.span) {
                 Some(IntRange(int_range))
             } else {
-                match (value.val, &value.ty.kind()) {
-                    (_, ty::Array(_, n)) => {
-                        let len = n.eval_usize(tcx, param_env);
-                        Some(Slice(Slice { array_len: Some(len), kind: FixedLen(len) }))
-                    }
-                    (ty::ConstKind::Value(ConstValue::Slice { start, end, .. }), ty::Slice(_)) => {
-                        let len = (end - start) as u64;
-                        Some(Slice(Slice { array_len: None, kind: FixedLen(len) }))
-                    }
-                    // FIXME(oli-obk): implement `deref` for `ConstValue`
-                    // (ty::ConstKind::Value(ConstValue::ByRef { .. }), ty::Slice(_)) => { ... }
-                    _ => Some(ConstantValue(value)),
-                }
+                Some(ConstantValue(value))
```

Comment:
Is this related to exhaustiveness checking? Shouldn't opaque consts be treated as matching nothing for that as they might not be structural-match?

Reply:
it is, very good catch. We still need a variant for string slices here, as string slices are not opaque, but `ConstantValue` is gone now and the replacement `Str` only handles `&str` constants. There are further improvements we can do here now that everything is explicit, but I'd rather not touch `compare_const_vals` in this PR.

Reply:
I should note that I understand basically nothing of this code, if I caught anything that was an accident.^^

Reply:
Also if the last 2 comments are bugfixes, is it possible to add a test?

Reply:
There was no bugfix, no behaviour was changed, all I did was to move the bail-out (that was guaranteed to happen later) to an earlier site so the rest of the code that I was able to remove became dead code. This PR generally is mostly just removing dead code that became obsolete after destructuring all structural-eq constants into (non-const) patterns. This means except for `&str`, all `PatKind::Const` are now opaque

Reply:
I see. Maybe add some doc comments to `PatKind` explaining this. :)